pa/biGraph.ml

```diff
@@ -0,0 +1,176 @@
+open Pervasives
+open BatPervasives
+open Apak
+
+include Log.Make(struct let name = "BiGraph" end)
+
+module type Vertex = sig
+  type t
+  val hash : t -> int
+  val equal : t -> t -> bool
+  val compare : t -> t -> int
+  val pp : Format.formatter -> t -> unit
+end
+
+module type S = sig
+  type t
+  type vertex
+
+  val pp : Format.formatter -> t -> unit
+
+  val empty : t
+  val make : vertex BatEnum.t -> vertex BatEnum.t -> (vertex * vertex) BatEnum.t -> t
+  val max_matching : t -> int
+  val u_size : t -> int
+  val v_size : t -> int
+  val e_size : t -> int
+end
+
+module Make (V : Vertex) = struct
+  type vertex = V.t
+  type edge = V.t * V.t [@@deriving ord]
+
+  module VertexSet = BatSet.Make(V)
+  module EdgeSet = BatSet.Make(struct
+      type t = edge [@@deriving ord]
+    end)
+
+  type t =
+    { u : VertexSet.t;
+      v : VertexSet.t;
+      e : EdgeSet.t }
+
+  let pp_edge formatter (u, v) =
+    Format.fprintf formatter "(%a, %a)" V.pp u V.pp v
+
+  let pp formatter graph =
+    Format.fprintf formatter "{U: {%a}, V: {%a}, E: {%a}}"
+    (ApakEnum.pp_print_enum V.pp) (VertexSet.enum graph.u)
+    (ApakEnum.pp_print_enum V.pp) (VertexSet.enum graph.v)
+    (ApakEnum.pp_print_enum pp_edge) (EdgeSet.enum graph.e)
+
+  let empty = 
+    { u = VertexSet.empty;
+      v = VertexSet.empty;
+      e = EdgeSet.empty }
+
+  let make u v e =
+    { u = VertexSet.of_enum u;
+      v = VertexSet.of_enum v;
+      e = EdgeSet.of_enum e}
+
+  module VMap = BatMap.Make(V)
+  module VoptMap = BatMap.Make(struct
```

Comment:
1. Is it important that None is the least element?  (If not we use [@@deriving ord] to implement the comparison).
2. I find the use of VoptMap to be a little confusing.  What's the advantage of using a VoptMap instead of a pair consisting of a VMap plus an extra element?